### ISSUE TITLE
Refactoring of Code Base

### ISSUE THREAD

**Issue by Noah11012** [DontBelieveMe/Stack#4]
Now that the error refactoring is complete, I think refactoring other parts of the code base should be done as well. Rename functions that are confusing to understand, refactoring the parser, and cleaning up the code in general. Adding better documentation is in order also.

I have one question about the `Macro` struct in parse.h. What does it do?

**Comment by DontBelieveMe** [comment]
now that is a very good question.
They where like my pretty naive attempt at implementing named variables (but they ended up more like simple text substitution - like C Macros) - basically I think you could "set" them using the `set` command (e.g `set value1 50`) and then anywhere where `value1` appeared from then on would be replaced with `50`
for example
```
set TEST 234
push TEST
```
would push 234 to the stack.
The struct itself 
```c
typedef struct _Macro {
	char *str_value;
	char *name;
} Macro;
```
just contains the name of the variable and the value of the variable in string form.

**Comment by Noah11012** [comment]
Does the text substitution happen before the actual parsing?

**Comment by DontBelieveMe** [comment]
it happens in `module_load` so thats before `module_execute` so it happens once at the start of the program and cannot be changed after that (if thats what you mean)

**Comment by Noah11012** [comment]
Ok, I got you.

**Comment by Noah11012** [comment]
What refactoring do you think we should do next?

**Comment by DontBelieveMe** [comment]
I think the most important thing (before anything else) is to write a new parser, starting with a tokenizer. Using an existing parser for this project is probably going to be overkill, so a simple recursive descent parser should be good enough.
Anyway but first a tokenizer is the order of the day. And ofc documenting as we go (instead of having to do it later).
How much do you know about lexing/parsing?

**Comment by Noah11012** [comment]
I know what lexing and parsing are but I'm not too sure of how to implement them in code. Right now I'm reading a series on how to create an interpreter at [Crafting Interpreters](http://www.craftinginterpreters.com/)

He goes over topics like lexing, parsering, grammar, recursive descent parsing, etc.

How do you think the lexer should be implemented? Scan across the entire file line by line breaking each line to a series of tokens that are stored in data structure? What data structure should it be? Linked list or a dynamic array?

A lot of decisions...

**Comment by Noah11012** [comment]
Now that I reread my previous comment I found how redundant it was. Lexing is the process of scanning the file breaking lines into tokens. What I should have said is what data structure we should use to hold the tokens?

**Comment by DontBelieveMe** [comment]
Well we can stream the tokens directly into the AST construction - something like
```c
struct Token get_next_token() {
    // .. parse next token
}

void parse_ast(struct Ast* ast) {
    while(!end_of_program) {
        struct Token tok = get_next_token();
       push_ast_node(ast, tok);
    }
}
```
(pseudocode) or something like that. That how I've done it in the past anyway. 
If not then we can just store the tokens in a simple list first and then parse the AST from that. A linked list should be good enough i think.

**Comment by Noah11012** [comment]
Should we add a "prepocessor"? If you want the ability to separate different instructions into different files we'll need something like it.

**Comment by DontBelieveMe** [comment]
I think a preprocessor would be cool - something similar to C's (but simpler if that makes sense, basic text subsitution, `#include` etc...)

**Comment by Noah11012** [comment]
Exactly what I was thinking. It should run before the parser (like in C) and then run the lexer and parser. So, I think we should first do the preprocessor first.

**Comment by DontBelieveMe** [comment]
yeah, that makes sense. I don't even think we'd need a tokenizer for the preprocessor - because the preprocessor only runs once, linearly (as in goes through the lines in order), so some simple regex or scanf's might be all that is needed. 

On a side note I think that it may be worth completly ditching the current `parse.c` file - as it's going to be changed almost completely anyway - refactoring it may turn out to be more hastle than it's worth.

**Comment by Noah11012** [comment]
I was thinking that, too. Parse.c is too much of a mess to fix. Needs to start from scratch.

Do you want to write the preprocessor or should I do it?

**Comment by DontBelieveMe** [comment]
you can have a go if you like - allthough i'm more happy to write or help with any bits.

**Comment by Noah11012** [comment]
Sure, I don't want to exclude you from the fun.

**Comment by DontBelieveMe** [comment]
😄  👍

**Comment by Noah11012** [comment]
Another important thing: what should be the "namespace" of the library? Many times libraries will prepend something to their functions to prevent name clashing with another library.

For example: if we have the function `get_next_token()` that is too generic and we should add something to mark it as ours. Like `stacky_get_next_token()`.

**Comment by DontBelieveMe** [comment]
This leads on to the fact that the project really needs a proper name - stack is just too generic. Stacky might be good enough for now?

**Comment by Noah11012** [comment]
It will work however we will get functions like `stacky_stack_new()`. Haha.

**Comment by Noah11012** [comment]
I think we should store each token in a linked list then create an AST. Is that overkill? Right now I'm reimplementing the linked list structure.

**Comment by Noah11012** [comment]
By the way, some of the initial refactoring is pushed on my fork so if you want to take a look go over to my fork.

**Comment by DontBelieveMe** [comment]
👍 I think storing each token in a linked list is OK for now, then we can create the AST from that. Tbh because it's a simple language (w/ simple syntax) generating an AST shouldn't be that complex.

**Comment by Noah11012** [comment]
Another design change is that the preprocessor will run along side the parser. If it encounters something that the PP should take care of, it will give it control and when its done it will continue to parse.

**Comment by DontBelieveMe** [comment]
So when the parser encounters a `#` followed by a `define` (for example), control will then be handed off to the preprocessor, which will then go over the whole document performing the text subsitution?

**Comment by Noah11012** [comment]
No. When the parser encounters an "instruction" like `include` or `set` it handles that case individually.

**Comment by Noah11012** [comment]
I'm refactoring the parser and I'm not sure what's the best way to tackle how the instructions should be parsed. Creating an AST for this would be in my opinion overkill. In the scanner function `stacky_lexer_begin()` splits the tokens based on a space or a tab. The parse function `stacky_parse_begin()` takes these tokens and depending on the instruction it executes the correct `stacky_instruction_*()` function. For example, if the `add` instruction is found, it executes `stacky_instruction_add()`. A second way of doing it would be to create a `StackyList` of `StackyInstructions` which is a struct that contains a function pointer to a `stacky_instruction_*()` function and a field for the operand passed to the instruction. For example, the line `push 100` would create the `StackyInstruction` with the fields `stacky_instruction_add` for the function pointer and `100` for the operand. Not sure what way you would like it to be.
